File: 0003/solution.py
#https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        if n < 2:
            return n
        head = 0
        tail = 1
        ans = 1
        while tail < n:
            while tail < n and s[tail] not in s[head:tail]:
                tail += 1
            ans = max(ans,tail-head)
            if tail < n:
                head += s[head:tail].index(s[tail])+1
        return ans

File: 0003/solution.java
//https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0) return 0;
        HashMap<Character,Integer>map = new HashMap<Character,Integer>();
        int ans = 0;
        int left = 0;
        for(int i=0;i<s.length();i++){
            if(map.containsKey(s.charAt(i))){
                left = Math.max(left,map.get(s.charAt(i))+1);
            }
            map.put(s.charAt(i),i);
            ans = Math.max(ans,i-left+1);
        }
        return ans;
    }
}